WeakSet should count its heap allocations

justinmichaud opened the pull request:
#### 2031dd94bcbea5f9b3324bcfcf09f5e5771b49fd
<pre>
WeakSet should count its heap allocations
<a href="https://bugs.webkit.org/show_bug.cgi?id=295072">https://bugs.webkit.org/show_bug.cgi?id=295072</a>

Reviewed by Yusuke Suzuki.

Long running pages that do not allocate, but have a RAF callback
can see their memory usage slowly increase forever. This is caused
by WeakBlock allocations. Without other kinds of allocations,
the collector thinks that the footprint isn&apos;t growing.

We notify the collector about these allocations now. Notably, it is
very important that we do not GC while allocating a Weak&lt;T&gt;, because
we construct them all over the place in the middle of inserting
into WeakMap. `didAllocate` does not trigger GC on its own, so this
patch does not change this behaviour.

Canonical link: <a href="https://commits.webkit.org/297950@main">https://commits.webkit.org/297950@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cf83389a5ad514b51058a1b7c75f6315d83c4b7b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110016 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29675 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20107 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116038 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60264 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111979 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30353 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38262 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83653 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112964 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24206 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99081 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64095 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23581 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17466 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59834 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/102507 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93579 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17272 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118829 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/108570 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37056 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27694 "Found 1 new test failure: fast/mediastream/getDisplayMedia-frame-rate.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92624 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37428 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95348 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92448 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/24234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37430 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15167 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32938 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36950 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42421 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/132844 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36612 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35939 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention-2.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39952 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38321 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Jarred-Sumner commented on Source/JavaScriptCore/heap/WeakSetInlines.h:
```diff
@@ -41,6 +41,7 @@ inline WeakImpl* WeakSet::allocate(JSValue jsValue, WeakHandleOwner* weakHandleO
     weakSet.m_allocator = allocator->next;
 
     WeakImpl* weakImpl = WeakBlock::asWeakImpl(allocator);
+    container.vm().heap.didAllocate(sizeof(WeakImpl));
```
Maybe didAllocateInBlock needs to be called to ensure the weak impl doesn't continue to appear in the weak block? (i'm just guessing at why its all red)

https://github.com/WebKit/WebKit/blob/f1f86afb2e78feb7393d53dd8fdfd901e06038c3/Source/JavaScriptCore/heap/LocalAllocator.cpp#L255

https://github.com/WebKit/WebKit/blob/f1f86afb2e78feb7393d53dd8fdfd901e06038c3/Source/JavaScriptCore/heap/MarkedSpace.cpp#L557-L563

justinmichaud replied:
Maybe I am reading this wrong. My understanding is that didAllocateInBlock is used to make sure that weak sets get updated. That is, if I allocate Foo, then didAllocateInBlock finds any WeakSets containing Foo, and clears them. This is allocating the WeakSet itself, so we don't need didAllocateInBlock. didAllocate is just for tracking the number of newly allocated bytes, but since this file is inlinable, it can be included from other compilation units, and thus we need to export it. Let's see what a second round of EWS thinks, to see if the failures were a random fluke or caused by this patch.

justinmichaud committed:
WeakSet should count its heap allocations
https://bugs.webkit.org/show_bug.cgi?id=295072

Reviewed by Yusuke Suzuki.

Long running pages that do not allocate, but have a RAF callback
can see their memory usage slowly increase forever. This is caused
by WeakBlock allocations. Without other kinds of allocations,
the collector thinks that the footprint isn't growing.

We notify the collector about these allocations now. Notably, it is
very important that we do not GC while allocating a Weak<T>, because
we construct them all over the place in the middle of inserting
into WeakMap. `didAllocate` does not trigger GC on its own, so this
patch does not change this behaviour.

Canonical link: https://commits.webkit.org/297950@main
